### ISSUE TITLE
history isn't shown if your system clock is set wrong

### ISSUE THREAD

**Issue by ridiculousfish** [fish-shell/fish-shell#6606]
fish wants to use a single history file, but also ignore item from other sessions that were ran after this session was started. To achieve this it compares the item's timestamp against the current clock. The problem is that if the clock is invalid or set backwards, then old items may appear new and will be ignored.

The real problem is that fish ever re-reads the file. What ought to happen is that fish uses whatever is in the file at session startup or when `history merge` is run, and then ignores the timestamp.

**Comment by mqudsi** [comment]
But that implies that we will be forced to ingest all history (or at least all that we can ever reference in the session without merging) and load it into memory? A more reliable alternative to the timestamps is a strictly globally monotonic session id identifier, but a more reasonable approach might be to just remember the length of the file/the seek position at startup.

**Comment by ridiculousfish** [comment]
No, it does not need to read the file, it only has to `mmap` it with `MAP_PRIVATE`. The file will still be lazily paged-in on demand. The `open/lseek` approach would also work; in fact, it might be even better since it would avoid the `SIGBUS` signals if the mapped file is truncated.

**Comment by scosu** [comment]
I am trying to use fish on embedded systems for development. Often the clock isn't set at all so fish doesn't offer any history to me after a fresh boot, even though the history file exists. I think it is the same issue as this one.
